[WIP] Extending AdaBoost's base estimator's feature_importances_ attr to support coef_ attr of an estimator

dsprahul opened the pull request:
Modified `feature_importances_` attribute in AdaBoost's base estimator to work with `coef_` attributes of given estimators if they don't provide `feature_importances_`  
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #12137 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
`feature_importances_` attribute of `BaseWeightBoosting` class now converts `coef_` attribute of input classifiers to their respective `feature_importances_`  using existing `_get_feature_importances` method in `feature_selection/from_model.py`.

This method accepts an optional argument: `norm_order` which defaults to `2` (`L2 norm`).
```
norm_order : {non-zero int, inf, -inf, "fro"}, optional
        Same as `numpy.linalg.norm`'s norm_order parameter
```   

#### Any other comments?
This method is now renamed to `get_feature_importances` and moved to `utils/__init__.py` as it is a shared resource.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

dsprahul committed:
Redesigned feature_importances_ attribute in AdaBoost's base estimator to accept "norm_type" parameter for
converting `coef_` to `feature_importances`.
Made a basic version of _coef_to_feature_importances function to handle this conversion

dsprahul committed:
missed comma

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging e2d069b18c42510f37f2e9549f83fd71be0c5830 into 63e5ae6319285c44b1a9519aa6acd9006c82a10a - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-8b160c21166dce031e07060770f20c2b69500cc2)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

dsprahul committed:
Missed renaming variables bug fix. doc skip

dsprahul committed:
Shortened a variable name to comply with <80c/l limit. doc skip

jnothman commented on sklearn/ensemble/weight_boosting.py:
```diff
@@ -234,11 +234,47 @@ def staged_score(self, X, y, sample_weight=None):
+            )
+        norm = non_negative_coefs_.sum()
+
+        return non_negative_coefs_ / norm
```
If the problem is multiclass, this return value will still, as far as I can tell, be 2-dimensional. `feature_importances_` must be 1d. I had intended that the norm would be used to aggregate coefficients across classes, as is done in [`_get_feature_importances`](https://github.com/scikit-learn/scikit-learn/blob/5fd9e03cf3a2e58d3b76dfaf0692437a4d6e1eec/sklearn/feature_selection/from_model.py#L15).

RahulDamineni replied:
Alright, I will reuse `_get_feature_importances` function from `feature_selection` to extend `feature_importances_` attribute of AdaBoost base class. Thanks for the link.

dsprahul committed:
Used `_get_feature_importances` from `feature_selection/from_model.py` to support estimators with `coef_`
parameter while calculating `feature_importances_` in AdaBoost based estimators. doc skip

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 61d9811fe28cbb41a05379ee2747a2921f080728 into 5fd9e03cf3a2e58d3b76dfaf0692437a4d6e1eec - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-fa80dad28c8eac32fcd761c347276d0e995c2ca2)

**new alerts:**

* 1 for Encoding error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

dsprahul committed:
Fixes encoding issue

jnothman commented on sklearn/ensemble/weight_boosting.py:
```diff
@@ -33,6 +33,7 @@
 from ..externals.six.moves import zip
 from ..externals.six.moves import xrange as range
 from .forest import BaseForest
+from ..feature_selection.from_model import _get_feature_importances
```
If it is a shared utility, it should live in sklearn.utils

jnothman replied:
The implementation currently mentions SelectFromModel and should not

RahulDamineni replied:
I see two solutions:
1. Move `_get_feature_importances` to `sklearn.utils.__init__.py` or create a new file (`feature_importances.py`) within `utils` and move it there; import this function into all those modules it's being used (2 places)
2. Without reusing this function, duplicate its functionality with-in `weight_boosting.py`; if the above step is unnecessary. 

@jnothman Which one should I choose?

jnothman reviewed (commented):
Thanks

jnothman commented:
Put it in utils/__init__.py for now. May also be able to reuse in RFE (I think there may be an issue somewhere about it only supporting one norm?)

dsprahul committed:
Moved `_get_feature_importances` to `utils/__init__.py` and renamed to `get_feature_importances`

RahulDamineni commented:
Alright, done. 
Yes, I think it's the issue #2121. I will look into it/ pass-on the update once this one is MGR

jnothman commented on sklearn/utils/__init__.py:
```diff
@@ -599,3 +599,39 @@ def is_scalar_nan(x):
     return bool(isinstance(x, (numbers.Real, np.floating)) and np.isnan(x))
+
+
+def get_feature_importances(estimator, norm_order=1):
```
we still might want to name this `_get_feature_importances`, otherwise we are promising API stability to users.

RahulDamineni replied:
Sure, will change it back to `_get_feature_importances`. Didn't think of that. 
Also, should I write more tests to pass codecov? (Currently it's 84.21%)

dsprahul committed:
renamed `get_feature_importances` to it's original

jnothman reviewed (commented):
Code coverage would be nice too, yes.

Sorry for all this messing about. It would be great to have an API for parameterized feature importances (I made an issue about it at some point), but it is not a small thing to change. I'm undecided on what to advise here.

RahulDamineni commented on sklearn/ensemble/weight_boosting.py:
```diff
@@ -234,10 +239,15 @@ def staged_score(self, X, y, sample_weight=None):
-    def feature_importances_(self):
-        """Return the feature importances (the higher, the more important the
-           feature).
+    def get_feature_importances_(self, norm_order=2):
```
It's okay. I will come-up with a solution by Monday.

RahulDamineni replied:
@jnothman: How about we make a Mixin for `feature_importances_` and inherit it where-ever it's required (Adaboost estimator, RFE, SelectFromModel etc..). This will require `norm_order` to be declared as a class variable in which-ever class it is being inherited into.

The job of this mixin is to determine whether a metaestimator or an usual estimator is inheriting it and then calculate `feature_importances_` accordingly. 

Something like this that can go into sklearn/base.py: 

```
class FeatureImportancesMixin(object):
    """Mixin class that generates ``feature_importances_`` attribute."""

    def _is_metaestimator(self):
        """Determines if `FeatureImportancesMixin` is inherited by an
        metaestimator.
        """
        return hasattr(self, "estimators_")

    def _get_feature_importances(estimator, norm_order):
        """Retrieve or aggregate feature importances from estimator.

        Parameters
        ----------
        estimator : any estimator object with either `feature_importances_`
            attribute or `coef_` attribute

        norm_order : {non-zero int, inf, -inf, "fro"}, optional
            Same as `numpy.linalg.norm`'s norm_order parameter

        Returns
        -------
        importances : A 1d array, shape: [n_features]
        """
        importances = getattr(estimator, "feature_importances_", None)

        coef_ = getattr(estimator, "coef_", None)
        if importances is None and coef_ is not None:
            if estimator.coef_.ndim == 1:
                importances = np.abs(coef_)

            else:
                importances = np.linalg.norm(coef_,
                                             axis=0,
                                             ord=self.norm_order)

        elif importances is None:
            raise AttributeError(
                "The underlying estimator %s has no `coef_` or "
                "`feature_importances_` attribute. Either pass a fitted estimator"
                " to SelectFromModel or call fit before calling transform."
                % estimator.__class__.__name__)

        return importances

    @property
    def feature_importances_(self):
        """Return the feature importances.

         The higher the value, the more important the feature.

        Returns
        -------
        feature_importances_ : array, shape = [n_features]
        """

           try:

                if self._is_metaestimator():
                    if self.estimators_ is None or len(self.estimators_) == 0:
                        raise ValueError(
                        "Estimator not fitted, "
                        "call `fit` before `feature_importances_`.")

                    norm = self.estimator_weights_.sum()
                    estimator_feature_importances_ = [
                        _get_feature_importances(estimator, self.norm_order)
                        for estimator in self.estimators_]

                    return (sum(weight * feature_importances_
                                for weight, feature_importances_ in zip(
                                    self.estimator_weights_,
                                    estimator_feature_importances_)) / norm)
                else:
                    return _get_feature_importances(estimator=self.estimator_,
                                                    norm_order=self.norm_order)

            except AttributeError:
                raise AttributeError(
                    "Unable to compute feature importances "
                    "since base estimator neither has a "
                    "feature_importances_ attribute nor a "
                    "coef_ attribute")
```

jnothman commented:
That design is not quite right, as I think you're confusing the meta estimator that uses the importances from the underlying estimator that provides only coef_.

Ultimately I think the norm is something users should be aware of when writing about how a model works, but is unlikely (?) something they would care to tweak.

This could be solved by deciding on one norm and sticking to it... Except that for a while we used different norms in SelectFromModel and RFE with no option to change. This is relatively hard to document... Although if we actually defined feature_importances_ on all estimators with coef_, that would be easy to document.

And if we went down the path of implementing get_importances() and deprecating feature_importances_, it would potentially involve adding an importances_params parameter to the meta estimators, but at least that would be clear and generic. 

I think for now we should just choose a norm. Yes, we could make it configurable through a class-level attribute, and yes we should reuse that function from SelectFromModel, but I now think that over-engineering the meta-estimator is the wrong thing to do when the design issue is in the constituent estimator.

dsprahul committed:
Made `norm_order` a class variable.
Wrote a test case to test the feature_importances_ calculation using coef_ attr of underlying estimator

RahulDamineni commented:
Alright, I made L2 norm as the default for AdaBoosts which could be configured via class-variable `norm_order`. I also added a test to test `coef_` to `feature_importances_` conversion.

hermidalc commented:
Actually @jnothman @RahulDamineni @alexsieusahai I don't think we should provide a parameter for the norm to make coef_ values non-negative, RFE and RFECV do not.  We should use the exact same code as what is there, at least as a first implementation: https://github.com/scikit-learn/scikit-learn/blob/bac89c253b35a8f1a3827389fbee0f5bebcbc985/sklearn/feature_selection/rfe.py#L181-L195

lorentzenchr commented:
See #12137.
